lib/profileAgents.js

```diff
@@ -602,6 +609,104 @@ async function _updateMeterController({meterId, controller}) {
+}
+
+async function _refreshProfileAgentZcaps({profileAgentRecords}) {
+  let zcapRefreshed = false;
+  const records = klona(profileAgentRecords);
+  const now = Date.now();
+  for(const record of records) {
+    if(!record.profileAgent.zcaps) {
+      continue;
+    }
+    const {zcaps} = record.profileAgent;
+    for(const zcapName in zcaps) {
+      if(zcapName !== 'profileCapabilityInvocationKey') {
+        const zcap = zcaps[zcapName];
+        const {expires} = zcap;
+        // check if the zcap is close to expiry
+        const expTime = Date.parse(expires);
+        const {zcapRefreshThreshold} = config.profile.profileAgent;
+        if(expTime - now <= zcapRefreshThreshold) {
+          if(zcapName === 'userDocument') {
+            await _refreshProfileAgentUserDocZcaps({zcap, now, record});
+          }
+          zcap.expires = new Date(now + DEFAULT_PROFILE_AGENT_ZCAP_TTL);
+          zcapRefreshed = true;
+        }
+      }
+    }
+    if(zcapRefreshed) {
+      // update the profile agent record
+      record.profileAgent.sequence += 1;
+      await update({profileAgent: record.profileAgent});
+    }
+  }
+  return zcapRefreshed;
+}
+
+async function _refreshProfileAgentUserDocZcaps({zcap, now, record} = {}) {
+  let zcapRefreshed = false;
+  const {zcapRefreshThreshold} = config.profile.profileAgent;
+  const kmsClient = new KmsClient({httpsAgent});
+  const profileSigner = await getProfileSigner({
+    kmsClient, profileAgentRecord: record
+  });
+  const docUrl = new URL(zcap.invocationTarget);
+  const edvId = `${docUrl.protocol}//${docUrl.hostname}:${docUrl.port}` +
+    `${docUrl.pathname.split('/').slice(0, 3).join('/')}`;
+  const edvClient = new EdvClient({id: edvId, httpsAgent});
+  const edvConfig = await getEdvConfig({edvClient, profileSigner});
+  const docId = zcap.invocationTarget.split('/').pop();
+  // get profile agent user doc
+  const profileAgentUserDoc = await getEdvDocument({
+    docId, edvConfig, edvClient, kmsClient, profileSigner
+  });
+  const updatedProfileAgentUserDoc = klona(profileAgentUserDoc);
+  const {zcaps: profileAgentUserDocZcaps} =
+    updatedProfileAgentUserDoc.content;
+  for(const profileAgentUserDocZcapName in profileAgentUserDocZcaps) {
+    if(profileAgentUserDocZcapName !== 'profileCapabilityInvocationKey') {
+      const profileAgentUserDocZcap =
+        profileAgentUserDocZcaps[profileAgentUserDocZcapName];
+      const {
+        expires: profileAgentUserDocZcapExpires
+      } = profileAgentUserDocZcap;
+      const profileAgentUserDocZcapExpTime =
+        Date.parse(profileAgentUserDocZcapExpires);
+      // check if the zcap is close to expiry
+      if(profileAgentUserDocZcapExpTime - now <= zcapRefreshThreshold) {
+        // refresh the zcap's expires property
+        profileAgentUserDocZcap.expires =
+          new Date(now + DEFAULT_PROFILE_AGENT_ZCAP_TTL);
+        zcapRefreshed = true;
+      }
+    }
+  }
+  if(zcapRefreshed) {
+    await edvClient.update({
```

Comment:
Needs to check for conflict / failure and retry until success or until there's no need to perform the update.

Reply:
Addressed [7c2b444](https://github.com/digitalbazaar/bedrock-profile/pull/99/commits/7c2b444cb7931466a50cd5fefc6589502a9ca36a)